Add dynamic Secondary Hero Shelf to desktop homepage

bobsilverberg opened the pull request:
Fixes #8249 

Screenshot:

![Screenshot 2019-09-10 09 01 03](https://user-images.githubusercontent.com/142755/64615986-92f56400-d3a9-11e9-9e7b-497095a0f2f0.png)

kumar303 commented on src/amo/components/SecondaryHero/styles.scss:
```diff
@@ -0,0 +1,63 @@
+
+.SecondaryHero-module-icon {
+  height: 60px;
+  margin: 16px;
```
Did you make adjustments to the spacing in the mocks? The spacing looks a little different, like the space between the icon and text and the overall padding.

mock:

<img width="362" alt="Screenshot 2019-09-10 16 14 58" src="https://user-images.githubusercontent.com/55398/64651262-ae329480-d3e6-11e9-8eca-9fd7c0e16cd8.png">


site:

<img width="343" alt="Screenshot 2019-09-10 16 14 43" src="https://user-images.githubusercontent.com/55398/64651264-b1c61b80-d3e6-11e9-9c62-178985e6f313.png">

bobsilverberg replied:
Honestly I was just trying to match the mock (and when I say mock I mean the Sketch file) as closely as possible using css. I can see that the icons don't look the same as in the mocks, but I'm using the size from the mock (60px) so I'm not sure what else I can do to match it.

I've gone back to Sketch to look at all the spacing between objects and have tried to match it more exactly. I was hoping to make a "best guess" and then ask @brassy- for feedback on any changes that were needed, but I have made some improvements from v1 of this patch.

kumar303 commented on src/amo/components/SecondaryHero/styles.scss:
```diff
@@ -0,0 +1,63 @@
+  line-height: 1.222;
+  margin: 16px;
+  padding: 0 16px 16px 16px;
+  position: relative;
```
What is this for? Is it overriding something?

kumar303 commented on tests/unit/amo/components/TestSecondaryHero.js:
```diff
@@ -0,0 +1,116 @@
+    expect(renderedModules).toHaveLength(3);
+
+    const renderedModule1 = renderedModules.at(0);
+    expect(renderedModule1.find('.SecondaryHero-module-icon')).toHaveProp(
```
Could all these assertions be done in an `it.each()` loop?

bobsilverberg commented on src/amo/components/SecondaryHero/styles.scss:
```diff
@@ -0,0 +1,63 @@
+  padding: 0 16px 16px 16px;
+  position: relative;
+  text-align: center;
+  width: 33%;
```
Yes, a grid is much better, thanks!

bobsilverberg commented on src/amo/components/SecondaryHero/styles.scss:
```diff
@@ -0,0 +1,63 @@
+}
+
+.SecondaryHero-message-headline {
+  font-size: 24px;
```
There is an existing var in `vars` that represents `24px`, but it's called `$font-size-form-header`. I originally used that, but then thought to myself "hey, self, this isn't a form header". Do you think I should just use `$font-size-form-header` here, or create a new var with a more generic name, like `$font-size-m-l`? 

Or maybe I should just use the existing `$font-size-l`, which is `28px` because that will make it more consistent with the rest of the site? `24px` is coming from the Sketch file, and I think one issue we might be hitting with these hero shelves is that the design specs have changed for Trailhead, so maybe we *shouldn't* be trying to be consistent with the rest of the site??

kumar303 replied:
For this initial launch I feel that we should try to keep the design consistent with the current AMO site. If you can use the existing variable for `28px` then that would be good. This type of subtle consistency is really important IMO. When we make all of this more trailhead-compliant we can hopefully address all of the historic inconsistencies that have crept in.

kumar303 commented on src/amo/components/SecondaryHero/styles.scss:
```diff
@@ -0,0 +1,72 @@
+  // justify-content: space-around;
+  line-height: 1.222;
+  // margin: 16px 0;
+  padding: 0 16px 16px 16px;
```
Where did you get `16px` from? The padding does not seem consistent in the mocks but it's mostly around `22px` or `23px`. What about making it `24px`? That would match the spacing of some nearby elements.

bobsilverberg replied:
I think `16px` was a best guess to try to make it look good. As you said, in the mock the padding is inconsistent. I have changed it to 24px as you suggest.

kumar303 commented on src/amo/components/SecondaryHero/styles.scss:
```diff
@@ -0,0 +1,72 @@
+}
+
+.SecondaryHero-message-description {
+  padding: $padding-page 0;
```
`$padding-page` is a weird value that crept in from long ago. I recommend avoiding it if you can. Our style guideline is to prefer multiples of 4 so I suggest `8px`. I see `9px` in the mock which is why I chose that.

Additionally, the bottom padding is `20px` in the mocks so I suggest making it `24px` for styleguide consistency.

kumar303 commented on src/amo/components/SecondaryHero/styles.scss:
```diff
@@ -0,0 +1,72 @@
+}
+
+.SecondaryHero-message {
+  font-size: 16px;
```
This can use `$font-size-m`

kumar303 commented on src/amo/components/SecondaryHero/styles.scss:
```diff
@@ -0,0 +1,72 @@
+  display: flex;
+  flex-direction: column;
+  // flex-grow: 0;
+  font-size: 18px;
```
Could this drop down to `$font-size-m` (16px) or is that too small?

bobsilverberg replied:
It could. The mock has this set at `18`, which is why I made it that. The mock shows that this is slightly larger than the font used for the description for the left-side message (`SecondaryHero-message`), which is why I used `18`. I feel like we're starting to play a bit fast and loose with the mocks the way we're changing a lot of the sizes, but I think it's probably okay as long as we keep them relatively the same. To change this so that it's the same size as another element in the mock, which the mock defines as being smaller than this, seems like maybe going too far.

kumar303 replied:
OK. I suggest declaring a new font variable for `18px` just so we have all the sizes in the same place. It could be something super specific like `$font-size-hero-description` or whatever.

bobsilverberg replied:
Interestingly, there is a `$font-size-m` defined as `18px` in `ui/css/vars.scss`, but I guess those values aren't used for components in the `amo` folder. I'll just create a specific one for now, as you suggest.

kumar303 commented on src/amo/components/SecondaryHero/styles.scss:
```diff
@@ -0,0 +1,72 @@
+  // margin: 16px 0;
+  padding: 0 16px 16px 16px;
+  // position: relative;
+  text-align: center;
```
I think you can remove this now that you have `align-items`, right?

bobsilverberg replied:
I can, but only if I move the rule into a selector for `SecondaryHero-module-description`, because it was actually needed for this. So I've done that as that's more explicit.

bobsilverberg commented on src/amo/components/SecondaryHero/styles.scss:
```diff
@@ -0,0 +1,72 @@
+  display: flex;
+  justify-content: space-between;
+  min-height: 216px;
+  padding: $padding-page-l 0;
```
I wasn't totally sure if you were saying we should have the gaps both above and below this shelf be 48px, or both be 24px. I am assuming the former and I'm making that change.

bobsilverberg commented on src/amo/components/SecondaryHero/index.js:
```diff
@@ -0,0 +1,57 @@
+  return (
+    <section className="SecondaryHero">
+      <div className="SecondaryHero-message">
+        <div className="SecondaryHero-message-headline">{headline}</div>
```
I had it as an `<h2>` originally, but that added a bunch of styles. I changed it back to an `<h2>` and explicitly overrode the styles I didn't want to inherit.

bobsilverberg commented on src/amo/components/SecondaryHero/styles.scss:
```diff
@@ -0,0 +1,72 @@
+  width: 25%;
+}
+
+.SecondaryHero-message-headline {
```
In the Sketch file, this is *not* marked as a bold font. It's marked as `Regular`, and I don't think it looks bold either. It just looks larger:

<img width="1021" alt="Screenshot 2019-09-16 14 51 06" src="https://user-images.githubusercontent.com/142755/64984774-78335b80-d891-11e9-9732-c939a4bd0840.png">

kumar303 replied:
Heh. I see now that Page 2 of the Sketch file has it in a regular font but Page 1 has it in a bold font! Here's Page 1 showing it in `FirefoxSharpSans-Bold`.

<img width="834" alt="Screenshot 2019-09-16 14 14 29" src="https://user-images.githubusercontent.com/55398/64986345-6ac7a280-d88c-11e9-9ec6-c5b60ebf7377.png">

bobsilverberg replied:
There's 2 pages?!?! That complicates things. :-(

bobsilverberg commented:
Thanks again for the review, @kumar303. I have addressed all of your comments except for two of them, both of which I question. Perhaps you can take a look and let me know what you think. I think ultimately we need to have @brassy- take a look at what we've come up with and let us know what, if anything, needs to change. That might be easiest done by landing an initial version on dev and letting her explore it there.

bobsilverberg commented:
@kumar303 I'm still waiting to connect with @brassy- on this one, but do you think it's okay to land on dev as it now is, and then we can make follow-up style changes if requested by @brassy-?

kumar303 commented:
Sorry, I didn't get a chance to re-review this morning. I can take a quick look right now because last time I thought there were unambiguous aspects to the mock that I wanted to see addressed. For anything ambiguous, yes, we could easily ask UX for more feedback.

bobsilverberg commented:
> Sorry, I didn't get a chance to re-review this morning. I can take a quick look right now because last time I thought there were unambiguous aspects of the mock that I wanted to see matched. For anything ambiguous, yes, we could easily ask UX for more feedback on dev.

I believe I have addressed all of the unambiguous issues you raised. This doesn't need to land for today's tag as it won't be active on stage or prod anyway.

kumar303 commented on src/amo/components/SecondaryHero/styles.scss:
```diff
@@ -0,0 +1,65 @@
+  color: $black;
+  display: grid;
+  grid-gap: 32px;
+  grid-template-columns: repeat(4, 1fr);
```
Even though we're not supporting this layout on Android, it's possible for Android users to see this screen from a deeplink that has `firefox` in the URL (see https://github.com/mozilla/addons-frontend/issues/6136). I whipped up a simple responsive view (only took a minute) which I think is worth including:

```diff
diff --git a/src/amo/components/SecondaryHero/styles.scss b/src/amo/components/SecondaryHero/styles.scss
index 055cd935c..c35f98250 100644
--- a/src/amo/components/SecondaryHero/styles.scss
+++ b/src/amo/components/SecondaryHero/styles.scss
@@ -3,10 +3,16 @@
 .SecondaryHero {
   color: $black;
   display: grid;
-  grid-gap: 32px;
-  grid-template-columns: repeat(4, 1fr);
+  grid-gap: $padding-page;
+  grid-template-columns: auto;
   min-height: 216px;
-  padding: 48px 0 24px 0;
+  padding: $padding-page 0 0 0;
+
+  @include respond-to(large) {
+    grid-gap: 32px;
+    grid-template-columns: repeat(4, 1fr);
+    padding: 48px 0 24px 0;
+  }
 }
 
 .SecondaryHero-message {
@@ -23,7 +29,11 @@
 }
 
 .SecondaryHero-message-description {
-  padding: 8px 0 24px 0;
+  padding: $padding-page 0;
+
+  @include respond-to(large) {
+    padding: 8px 0 24px 0;
+  }
 }
 
 .SecondaryHero-module {
@@ -34,13 +44,23 @@
   flex-direction: column;
   font-size: $font-size-hero-description;
   line-height: 1.222;
-  padding: 0 24px 24px 24px;
+  padding: $padding-page;
+  padding-top: 0;
+
+  @include respond-to(large) {
+    padding: 24px;
+    padding-top: 0;
+  }
 }
 
 .SecondaryHero-module-icon {
   height: 60px;
-  margin: 30px auto 36px auto;
+  margin: $padding-page auto;
   width: 60px;
+
+  @include respond-to(large) {
+    margin: 30px auto 36px auto;
+  }
 }
 
 .SecondaryHero-module-link {
```

bobsilverberg replied:
Thanks! I agree that looks better on a small sceen!

kumar303 commented on src/amo/components/SecondaryHero/styles.scss:
```diff
@@ -0,0 +1,65 @@
+  display: grid;
+  grid-gap: 32px;
+  grid-template-columns: repeat(4, 1fr);
+  min-height: 216px;
```
What were you trying to address with this? If I remove this rule, the current dev home page looks the same.

bobsilverberg replied:
I think I did this when I was playing around with flexbox, but now with the grid it does not seem to be needed anymore.

willdurand commented:
@bobsilverberg merging this patch broke all the circle-ci tests. It could be because there is no data for the hero shelves in the addons-server docker env, but I am not sure. In any case, it'd be nice to get this fixed. cc @jrbenny35
